### PR TITLE
Add default_trait_access lint

### PR DESCRIPTION
**Pull request by illicitonion**

### CODE REVIEW

**Comment by mati865** [comment]
New lint found warnings in old tests.

**Comment by illicitonion** [comment]
Looks like I still have a little more debugging to do around derived `Default` implementations :)

**Comment by illicitonion** [comment]
Ok, this should be good to review now - there's one TODO for which any pointers would be appreciated!

File: clippy_lints/src/default_trait_access.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    fn get_lints(&self) -> LintArray {
+        lint_array!(DEFAULT_TRAIT_ACCESS)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for DefaultTraitAccess {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if_chain! {
+            if let ExprCall(ref path, ..) = expr.node;
+            if !any_parent_is_automatically_derived(cx.tcx, expr.id);
```
Comment (oli-obk) [inline clippy_lints/src/default_trait_access.rs]: You found a workaround for the currently broken `in_macro`! Sweet.

@rust-lang-nursery/clippy  do we want to add this to the in_macro check? It's probably expensive to do, but we could monoize it via a thread local HashMap

Comment (oli-obk) [reply clippy_lints/src/default_trait_access.rs]: I should have formulated this better:

`in_macro` is broken and does not help us anymore in detecting `derive`s. You just found a way to detect derives again, and we should make `in_macro` also call your function in order to make it work again.

Comment (illicitonion) [reply clippy_lints/src/default_trait_access.rs]: Fabulous! :D

**Comment by oli-obk** [comment]
Anyway, that discussion should not be blocking this PR.

**Comment by oli-obk** [comment]
In case you are looking for something to do, you could modify our `in_macro` function to additionally take a `cx` and a `NodeId`, and if the original function body would return `false`, also run your check.

You'd probably end up touching every single file in `clippy_lints`, so it is a little bit of effort. The fallout should be a bunch of ui tests suddenly getting fixed :D

**Comment by Manishearth** [comment]
Hey, should this be pedantic? This is a pretty common pattern.

**Comment by Manishearth** [comment]
Discussion moved to https://github.com/rust-lang-nursery/rust-clippy/pull/2872

**Pull request merged**